File: nodejs/src/service/userService.ts
import { FriendNode, User } from "../types/general";
import repository from "../repository/userFriendsRepository";

const friendTreeBFS = (startId: string, maxDepth: number): FriendNode => {
  const rootUser = repository.getUserById(startId) as User;

  const rootNode: FriendNode = {
    id: rootUser.id,
    name: rootUser.name,
    friends: [],
  };

  const queue: Array<{ node: FriendNode; depth: number }> = [
    { node: rootNode, depth: 0 },
  ];

  const visited = new Set<string>();
  visited.add(startId);

  while (queue.length > 0) {
    const { node, depth } = queue.shift()!;

    if (depth >= maxDepth) continue;

    const friends = repository.getFriendsById(node.id);

    for (const friend of friends) {
      if (visited.has(friend.id)) continue;

      visited.add(friend.id);

      const child: FriendNode = {
        id: friend.id,
        name: friend.name,
        friends: [],
      };

      node.friends.push(child);
      queue.push({ node: child, depth: depth + 1 });
    }
  }

  return rootNode;
};

export default friendTreeBFS;

File: kotlin/friends-tree/src/main/kotlin/com/pret/friend/tree/service/UserService.kt
package com.pret.friend.tree.service

import com.pret.friend.tree.abstractions.UserNode
import com.pret.friend.tree.repository.UserFriendsRepository
import org.springframework.stereotype.Service

@Service
class UserService(
    private val userFriendsRepository: UserFriendsRepository
) {

    fun friendTreeBfs(rootUserId: Int, maxDepth: Int): UserNode {
        val rootUser = userFriendsRepository.getUserById(rootUserId)
        val rootNode = UserNode(rootUser.id, rootUser.name)

        val visited = mutableSetOf(rootUserId)
        val queue: ArrayDeque<Triple<UserNode, Int, Int>> = ArrayDeque()

        queue.add(Triple(rootNode, rootUserId, 0))

        while (queue.isNotEmpty()) {
            val (parentNode, currentUserId, depth) = queue.removeFirst()

            if (depth == maxDepth) continue

            val friends = userFriendsRepository.getFriendsById(currentUserId)

            for (friend in friends) {
                if (friend.id !in visited) {
                    visited.add(friend.id)

                    val childNode = UserNode(friend.id, friend.name)
                    parentNode.friends.add(childNode)

                    queue.add(Triple(childNode, friend.id, depth + 1))
                }
            }
        }

        return rootNode
    }

}